Track spawn `Tick` of entities, offer methods, query data `SpawnDetails` and query filter `Spawned`

urben1680 opened the pull request:
# Objective

In my own project I was encountering the issue to find out which entities were spawned after applying commands. I began maintaining a vector of all entities with generational information before and after applying the command and diffing it. This was awfully complicated though and has no constant complexity but grows with the number of entities.

## Solution

Looking at `EntyMeta` it seemed obvious to me that struct can track the tick just as it does with `MaybeLocation`, updated from the same call. After that it became almost a given to also introduce query data `SpawnDetails` which offers methods to get the spawn tick and location, and query filter `Spawned` that filters entities out that were not spawned since the last run.

## Testing

I expanded a few tests and added new ones, though maybe I forgot a group of tests that should be extended too. I basically searched `bevy_ecs` for mentions of `Changed` and `Added` to see where the tests and docs are.

Benchmarks of spawn/despawn can be found [here](https://github.com/bevyengine/bevy/pull/19047#issuecomment-2852181374).

---

## Showcase

From the added docs, systems with equal complexity since the filter is not archetypal:
```rs
fn system1(q: Query<Entity, Spawned>) {
    for entity in &q { /* entity spawned */ }
}

fn system2(query: Query<(Entity, SpawnDetails)>) {
    for (entity, spawned) in &query {
        if spawned.is_spawned() { /* entity spawned */ }
    }
}
```

`SpawnedDetails` has a few more methods:

```rs
fn print_spawn_details(query: Query<(Entity, SpawnDetails)>) {
    for (entity, spawn_details) in &query {
        if spawn_details.is_spawned() {
            print!("new ");
        }
        println!(
            "entity {:?} spawned at {:?} by {:?}",
            entity,
            spawn_details.spawned_at(),
            spawn_details.spawned_by()
        );        
    }
}
```

## Changes

No public api was changed, I only added to it. That is why I added no migration guide.

- query data `SpawnDetails`
- query filter `Spawned`
- method `Entities::entity_get_spawned_or_despawned_at`
- method `EntityRef::spawned_at`
- method `EntityMut::spawned_at`
- method `EntityWorldMut::spawned_at`
- method `UnsafeEntityCell::spawned_at`
- method `FilteredEntityRef::spawned_at`
- method `FilteredEntityMut::spawned_at`
- method `EntityRefExcept::spawned_at`
- method `EntityMutExcept::spawned_at`

alice-i-cecile commented:
I really like this idea: it's useful and consistent. That said, we should see how this affects spawning performance, both in the worst case and on more realistic cases.

ElliottjPierce commented on crates/bevy_ecs/src/query/filter.rs:
```diff
@@ -1005,6 +1007,145 @@ unsafe impl<T: Component> QueryFilter for Changed<T> {
+/// # Time complexity
+///
+/// `Spawned` is not [`ArchetypeFilter`], which practically means that if query matches million
+/// entities, `Spawned` filter will iterate over all of them even if none of them were spawned.
```
This is true, but kinda disappointing. I don't know if we could reasonably do better though. Probably not.

Like, we could maintain a list of entities, sorted by their spawn time, then just iterate until we hit one that was spawned too late. But that's terrible cache locality. Not to mention despawn time.

urben1680 replied:
A sorted vector sounds good in theory if iterating the query was random access from a list of entities.
But instead they match archetypes and iterate through it's table rows where the filter checks each row for the ticks if the filter is not archetypal, or just iterate the rows without further checking if the filter is archetypal.

So I don't see how such a sorted vector could be a benefit.

urben1680 replied:
What might work better if the table rows were sorted by the time the entity was spawned. But implementing that is beyond what I am comfortable to implement.

ElliottjPierce replied:
Yeah, I don't think we can do better than what you have. Just wish we could.

ElliottjPierce commented on crates/bevy_ecs/src/query/fetch.rs:
````diff
@@ -470,6 +472,118 @@ unsafe impl QueryData for EntityLocation {
+/// # bevy_ecs::system::assert_is_system(print_spawn_ticks);
+/// ```
+#[derive(Clone, Copy)]
+pub struct SpawnedTick(());
````
Same thing here. Why not this?
```suggestion
pub struct SpawnedTick;
```

ElliottjPierce commented on crates/bevy_ecs/src/entity/mod.rs:
```diff
@@ -1025,16 +1024,41 @@ impl Entities {
+    }
+
+    #[inline]
+    pub(crate) fn check_change_ticks(&mut self, change_tick: Tick) {
```
Where is this called? Is this meant to advance the change tick to help with wrapping? IDK. Maybe a doc comment would be helpful too.

urben1680 replied:
It is called in the this added line within the `World::check_change_ticks` method:
https://github.com/bevyengine/bevy/pull/19047/files#diff-8799631eaab5e61e5dbc20251e08d83474df566ba159061cf2a78ce8f1fd59d5R3149

You are correct this is needed to keep the wrapping working. If a spawn gets too long ago then `too_long_ago_spawn_tick.is_newer_than(last_run, this_run)` should always return `false`. This behavior is only consistentl if this method here is called before overflow would happen.

ElliottjPierce replied:
Ok. That makes sense. I just haven't messed with ticks before so it was confusing. But I think you're good here then!

ElliottjPierce commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2952,6 +2963,14 @@ impl<'w> EntityWorldMut<'w> {
+        self.world()
+            .entities()
+            .entity_get_spawned_or_despawned_at(self.entity)
+            .unwrap()
```
I don't like this unwrap (but it may be the best choice). Do you think we could justify an unchecked unwrap (or an unchecked method on `Entities`) since the `EntityWorldMut` is always valid? Or if not, could we make this a result/option?

urben1680 replied:
This pretty much works like the method right above, I simply copied the same access. I lack the confidence to make this more risky with an unchecked unwrap but since this method here may be hotter than the other, it is an option. But I would like to postpone this to benchmarks.

urben1680 replied:
Okay I decided to make an extra unsafe method because it really should be unable to fail as the queried entity must be alive here.

ElliottjPierce commented on crates/bevy_ecs/src/world/unsafe_world_cell.rs:
```diff
@@ -1136,6 +1136,14 @@ impl<'w> UnsafeEntityCell<'w> {
+        self.world()
+            .entities()
+            .entity_get_spawned_or_despawned_at(self.entity)
+            .unwrap()
```
Same goes here. Is unwrap the best choice?

ElliottjPierce reviewed (commented):
Over all, looks good. Just some minor questions, etc.

One thing we should consider is benching this. I have a feeling that having a separate vec for `SpawnedOrDespawned`, separate from `EntityMeta` would be faster. Almost every iteration cares about where the entity is. Asking when it was spawned and by who is more like a map thing IIUC. But this might be better as a separate PR.

Another option is making an observer and event for `Spawned`. I feel like that's going to be the primary use of this, and exposing an observer might be better anyway for this use case. Then again, it makes sense to ask when an entity was spawned. Ex: This particle entity has been around for 100 ticks; despawn it. (Not that each particle should be an entity.) In other words, this might be a cheaper alternative to a timer component in some cases.

Anyway, that's my thoughts. I'm generally on board with this, provided the separate vecs and observer APIs are investigated (maybe in future prs).

urben1680 commented on crates/bevy_ecs/src/query/filter.rs:
````diff
@@ -1005,6 +1007,145 @@ unsafe impl<T: Component> QueryFilter for Changed<T> {
+///
+/// # bevy_ecs::system::assert_is_system(print_spawning_entities);
+/// ```
+pub struct Spawned(());
````
This prevents it to be manually constructed, since this is only used as a type and not as a value. But there is no harm if one would construct it somewhere so I guess I can make that a simple field-less struct.

urben1680 commented:
Regarding observers, I considered it but I think the PR is big enough as is. I also imagine spawns triggering observers will have a more noticeable performance impact on all spawns than what I did here.

urben1680 commented:
I did another change.

Previously, `Entities::set_spawned_or_despawned_by` was called at the end of spawns/despawns, which was fine for logging something you usually deactivate during releases.

However, since with this change the method is no longer noop even when you deactivate tracking locations, this always added to the number of accesses on `Entities`.

So I removed that method entirely and added `Entities::set_spawn_despawn` that is used _instead of_ `Entities::set` in the spawning context.

`Entities::set` is still used at
- `BundleInserter::insert`
- `EntityWorldMut::move_entity_from_remove` (which makes me think #18521 does not need to adjust here)

`Entities::set_spawn_despawn` is now used instead at
- `BundleSpawner::spawn_non_existent`
- `EntityWorldMut::despawn_with_caller` (I don't know why this method has two set calls, please tell me)
- `World::spawn_at_empty_internal`

When doing spawn benchmarks, this change greatly reduced the regressions to +/-1%.
Despawn benchmarks are still a bit wildly going in +/- 5%.

I will do a separate post with benchmarks once I am sure I do them correctly.

alice-i-cecile commented:
The other idea for this that I had is that we could add a permanent marker component to every entity on first spawn. In some ways that would be more intrusive (shows up in inspectors), but in other ways it would be less so, since we could reuse existing machinery. I *think* I prefer this style of solution (but splitting this out from `EntityMeta` for perf reasons makes sense), but I wanted to raise it as an option.

ElliottjPierce commented:
Also, as a principal, I think any immutable component that every entity needs should, by definition, be stored in `Entities`. It's kinda a base table. The only question is how to split up the vecs. (And yeah, definitely worth benching that. I have no idea which way will be faster.)

urben1680 commented:
I implemented the idea of having a separate vector.

Instead of an `Option<Tick>` I went for an `MaybeUninit<Tick>` because the first doubles the type size. The checked read is done by using the `EntityMeta` stuff. We can consider doing the same with the location but for simplicity I kept that as it was before the PR, besides the different way to set them.

I will tomorrow or so do the benchmarks on the last commit and the one with the previous implementation to see which is better.

chescock reviewed (approved):
This is a really cool idea!

urben1680 commented on crates/bevy_ecs/src/query/fetch.rs:
```diff
@@ -470,6 +472,118 @@ unsafe impl QueryData for EntityLocation {
+///
+/// fn print_spawn_ticks(query: Query<(Entity, SpawnedTick)>, system_ticks: SystemChangeTick) {
+///     for (entity, spawned) in &query {
+///         if spawned.is_newer_than(system_ticks.last_run(), system_ticks.this_run()) {
```
At first I had a more complex type for the item, but I disliked the `spawned_tick.spawned_tick()` hiccup you would end up with. Though if the query data was something like `SpawnDetails` then it'd be a nicer `details.spawned_tick()` / `details.is_spawned()` / `details.spawned_details()`.

Gonna integrate that idea in this PR since it is not that much more with a bit of mapping.

urben1680 commented on crates/bevy_ecs/src/entity/mod.rs:
```diff
@@ -686,6 +703,7 @@ impl Entities {
         } else {
             let index = u32::try_from(self.meta.len()).expect("too many entities");
             self.meta.push(EntityMeta::EMPTY);
+            self.spawned_or_despawned_at.push(MaybeUninit::uninit());
```
Right, though having the compiler not even write a "dummy value" here but uninit data could cut another nanosecond. 😆

urben1680 commented:
Here are the results of benchmarking on my Ryzen 7700:

- main: e05e74a76adea3c170f5bf8dafecd373a8ae6488
- in-entity-meta f27aa7803cd6aea4d0bb5a6bbd050fd4f06e7fd5
- separate-vec d094ae73ca3212308852ca181e0c362a3992de65

| group | in-entity-meta | main | separate-vec |
| - | - | - | - |
| despawn_world/10000_entities | $\color{lime}{\textsf{1.00 - 363.9±4.67µs}}$ | 1.01 - 369.1±4.49µs | 1.02 - 371.3±14.92µs |
| despawn_world/1000_entities | $\color{lime}{\textsf{1.00 - 37.3±0.40µs}}$ | 1.00 - 37.4±0.29µs | 1.01 - 37.6±0.60µs |
| despawn_world/100_entities | $\color{lime}{\textsf{1.00 - 4.4±0.11µs}}$ | 1.00 - 4.4±0.06µs | 1.01 - 4.5±0.06µs |
| despawn_world/10_entities | $\color{lime}{\textsf{1.00 - 665.1±15.16ns}}$ | 1.01 - 671.1±10.43ns | 1.02 - 678.2±10.23ns |
| despawn_world/1_entities | 1.00 - 145.7±6.38ns | $\color{lime}{\textsf{1.00 - 145.1±3.58ns}}$ | 1.04 - 150.4±4.09ns |
| despawn_world_recursive/10000_entities | 1.04 - 2.2±0.06ms | 1.01 - 2.2±0.03ms | $\color{lime}{\textsf{1.00 - 2.2±0.01ms}}$ |
| despawn_world_recursive/1000_entities | 1.04 - 219.0±3.72µs | $\color{lime}{\textsf{1.00 - 211.5±1.77µs}}$ | 1.02 - 216.4±8.16µs |
| despawn_world_recursive/100_entities | 1.01 - 22.8±0.47µs | $\color{lime}{\textsf{1.00 - 22.5±0.29µs}}$ | 1.00 - 22.6±0.20µs |
| despawn_world_recursive/10_entities | 1.03 - 2.7±0.06µs | $\color{lime}{\textsf{1.00 - 2.6±0.04µs}}$ | 1.02 - 2.7±0.05µs |
| despawn_world_recursive/1_entities | 1.01 - 397.5±11.12ns | $\color{lime}{\textsf{1.00 - 394.5±10.60ns}}$ | 1.02 - 401.1±12.32ns |
| spawn_world/10000_entities | $\color{lime}{\textsf{1.00 - 503.4±40.53µs}}$ | 1.02 - 512.2±48.34µs | 1.01 - 509.2±44.93µs |
| spawn_world/1000_entities | $\color{lime}{\textsf{1.00 - 50.2±4.35µs}}$ | 1.02 - 51.5±5.05µs | 1.02 - 51.2±4.97µs |
| spawn_world/100_entities | $\color{lime}{\textsf{1.00 - 5.0±0.45µs}}$ | 1.02 - 5.1±0.52µs | 1.01 - 5.1±0.52µs |
| spawn_world/10_entities | $\color{lime}{\textsf{1.00 - 505.0±43.58ns}}$ | 1.01 - 509.5±54.93ns | 1.01 - 509.4±52.75ns |
| spawn_world/1_entities | 1.00 - 50.8±4.71ns | $\color{lime}{\textsf{1.00 - 50.6±5.18ns}}$ | 1.01 - 51.3±4.80ns |

After adding the same idea with `MaybeUninit` to `EntityMeta`, the first approach got better, for some reason even better than `main` except for recursive despawns.

If there are no objections I will bring the approach of having all in `EntityMeta` back into this branch.

@alice-i-cecile are these sufficient benchmarks for the tag to be removed?

alice-i-cecile commented:
> If there are no objections I will bring the approach of having all in EntityMeta back into this branch.

Sounds good! Thanks for benchmarking this!

ElliottjPierce commented on crates/bevy_ecs/src/entity/mod.rs:
```diff
@@ -1097,15 +1166,21 @@ struct EntityMeta {
     pub location: EntityLocation,
     /// Location of the last spawn or despawn of this entity
-    spawned_or_despawned_by: MaybeLocation<Option<&'static Location<'static>>>,
+    spawned_or_despawned: MaybeUninit<SpawnedOrDespawned>,
```
I'm not opposed to introducing `MaybeUninit` here, but I want to draw attention to it. Is this worth it over a dummy value? Even a dummy values of all 0s?

urben1680 replied:
The alternative that makes the most sense to me would be to separate `Tick` with a dummy value + `MaybeLocation` with an `Option<Location>` inside. I guess that would be the same thing minus a bit of unsafe code.
What I prefer here is that is shows both are valid or both are invalid. But I can do it either way.

ElliottjPierce reviewed (approved):
Looks good to me. Benches look good too. Great idea.

alice-i-cecile commented on crates/bevy_ecs/src/query/filter.rs:
```diff
@@ -1062,6 +1064,146 @@ unsafe impl<T: Component> QueryFilter for Changed<T> {
+///
+/// Note, that entity spawns issued with [`Commands`](crate::system::Commands)
+/// are visible only after deferred operations are applied,
+/// typically at the end of the schedule iteration.
```
```suggestion
/// typically at the end of each schedule.
```

urben1680 replied:
This is wording I copied from `Changed` and `Added`, I think this can be something else entirely since I think it originates from before auto sync point insertion. Today it typically is after the system that issues the command.

alice-i-cecile reviewed (changes requested):
Awesome stuff: that's exactly the API I was hoping for and I'm happy with those benchmarks. Unfortunately, you've made this too cool, and now I'm going to make you write release notes ;)

NthTensor commented:
At minimum, you can probably just copy over your showcase section to a new file following the instructions posted above. 

This is super neat!

urben1680 commented:
I wrote release notes. I wanted to do it detailed, I hope not too much.

[rendered](https://github.com/bevyengine/bevy/blob/5dc96955353664d9474d63e6aa6b005f49ad50b2/release-content/release-notes/entity-spawn-ticks.md)

I was also close to give a short recap that `Tick`s becoming too old is something to keep in mind but I was not sure if this belongs there.

alice-i-cecile reviewed (approved):
Good stuff! Thanks for the release note; it'll need an editing pass but capturing the motivation and usage is the key thing at this stage.

alice-i-cecile commented:
Merge conflicts, sorry!

urben1680 commented:
@alice-i-cecile resolved them. :+1:

akimakinai commented on crates/bevy_ecs/src/world/entity_ref.rs:
```diff
@@ -2453,27 +2466,22 @@ impl<'w> EntityWorldMut<'w> {
-        unsafe {
-            world
-                .entities_mut()
-                .set_spawned_or_despawned_by(self.entity.index(), caller);
```
caller is no longer set for entities not `swap_remove`d? #21293

<sub>(sorry, I commented this on merge commit initially)</sub>
